types/react-query/index.d.ts

```diff
@@ -36,107 +53,201 @@ export interface QueryOptions<TResult> {
+    status: 'error';
+    resolvedData: undefined;
+    latestData: undefined;
+    error: Error;
+}
+
+export interface PaginatedQuerySuccessState<TResult> extends QueryBaseState {
+    status: 'success';
+    resolvedData: TResult;
+    latestData: TResult;
+    error: null;
+}
+
+export type InfiniteQueryState<TResult, TVariable> =
+    | InfiniteQuerySuccessState<TResult, TVariable>
+    | InfiniteQueryErrorState<TVariable>
+    | InfiniteQueryLoadingState<TVariable>;
+
+export interface InfiniteQueryLoadingState<TVariable> extends InfiniteQueryBaseState<TVariable> {
+    status: 'loading';
+    data: [];
+    error: null;
+}
+
+export interface InfiniteQueryErrorState<TVariable> extends InfiniteQueryBaseState<TVariable> {
+    status: 'error';
+    data: [];
+    error: Error;
+}
 
-export interface PrefetchQueryOptions<TResult> extends QueryOptions<TResult> {
-    force?: boolean;
+export interface InfiniteQuerySuccessState<TResult, TVariable> extends InfiniteQueryBaseState<TVariable> {
+    status: 'success';
+    data: TResult[];
+    error: null;
+}
+
+export interface InfiniteQueryBaseState<TVariable> {
+    isFetching: boolean;
+    isFetchingMore: boolean;
+    failureCount: number;
+    refetch: (options?: {
+        force?: boolean;
+        thrownOnError?: boolean;
+    }) => void;
+    fetchMore: (variables?: TVariable) => Promise<any>;
+    canFetchMore: boolean;
 }
 
 export function useMutation<TResults, TVariables extends object>(
     mutationFn: MutationFunction<TResults, TVariables>,
-    mutationOptions?: MutationOptions
-): [MutateFunction<TResults, TVariables>, MutationResult<TResults>];
+    mutationOptions?: MutationOptions<TResults>,
+): [MutateFunction<TResults, TVariables>, MutationState<TResults>];
 
-export type MutationFunction<TResults, TVariables extends object> = (
-    variables: TVariables,
-) => Promise<TResults>;
+export type MutationFunction<TResults, TVariables extends object> = (variables: TVariables) => Promise<TResults>;
 
-export interface MutationOptions {
-    refetchQueries?: Array<string | [string, object]>;
-    refetchQueriesOnFailure?: boolean;
+export interface MutationOptions<TResult> {
+    onSuccess?: (data: TResult) => Promise<any> | void;
+    onSettled?: (data: TResult, error: any) => Promise<any> | void;
+    onError?: (error: any) => Promise<any> | void;
+    throwOnError?: boolean;
+    useErrorBoundary?: boolean;
 }
 
 export type MutateFunction<TResults, TVariables extends object> = (
     variables?: TVariables,
     options?: {
-        updateQuery?: string | [string, object],
+        updateQuery?: string | [string, object];
         waitForRefetchQueries?: boolean;
-    }
+    },
 ) => Promise<TResults>;
 
-export interface MutationResult<TResults> {
-    data: TResults | null;
-    isLoading: boolean;
+export interface MutationState<TResults> {
+    status: 'idle' | 'loading' | 'error' | 'success';
+    data: TResults | undefined;
```

Comment:
missing `reset` function
https://github.com/tannerlinsley/react-query#resetting-mutation-state
```suggestion
    data: TResults | undefined;
    reset: () = void;
```

Reply:
Thanks. Its also worth raising an issue in react-query repo to fix the docs, which misses this `reset` function